### PR TITLE
18USA P10

### PR DESCRIPTION
**Pull request by ventusignis**
Implement 18USA P10: Carnegie Steel Company

### CODE REVIEW

**Comment by crericha** [comment]
FYI I'm working on P9. Almost have it done.

File: lib/engine/game/g_18_usa/step/buy_sell_par_shares.rb [hunk 1]
```diff
@@ -137,6 +137,16 @@ def process_assign(action)
               company.close!
             end
 
+            corporation_hex = corporation.tokens.first.hex
+
+            if company.id == 'P10' && @game.potential_metropolitan_hexes.include?(corporation_hex.id) &&
+                @game.active_metropolitan_hexes.none? { |hex| hex.id == corporation_hex.id }
```
Comment (crericha) [inline lib/engine/game/g_18_usa/step/buy_sell_par_shares.rb:143]: !@game.active_metropolitan_hexes.include?(corporation_hex)

File: lib/engine/game/g_18_usa/game.rb [hunk 2]
```diff
@@ -668,6 +668,13 @@ class Game < G1817::Game
         BRIDGE_CITY_HEXES = %w[I19 C10 C17 D14 E15 E17 F20 G17].freeze
         BRIDGE_TILE_HEXES = %w[B4 D18 E21 F18 H18].freeze
 
+        # Does not include guaranteed metropolis New York City
+        POTENTIAL_METROPOLIS_HEXES = %w[D20 E11 G3 H14 H22 I19].freeze
+
+        def potential_metropolitan_hexes
+          POTENTIAL_METROPOLIS_HEXES
```
Comment (crericha) [inline lib/engine/game/g_18_usa/game.rb:675]: Recommend converting to Hex objects here and caching the result.

File: lib/engine/game/g_18_usa/step/buy_sell_par_shares.rb [hunk 3]
```diff
@@ -137,6 +137,16 @@ def process_assign(action)
...
+              @game.log << "#{company.name} turns #{corporation_hex.location_name} into a metropolis"
+              puts 'metttttro'
```
Comment (crericha) [inline lib/engine/game/g_18_usa/step/buy_sell_par_shares.rb:145]: Remove debug

File: lib/engine/game/g_18_usa/game.rb [hunk 4]
```diff
@@ -1605,32 +1628,34 @@ def setup
...
-            end
-          end
+          metro_hexes = METROPOLITAN_HEXES.sort_by { rand }.take(3)
+          metro_hexes.each { |metro_hex| convert_potential_metro(metro_hex) }
 
           randomize_subsidies
         end
 
+        # Convert a potential metro hex to a metro hex
+        def convert_potential_metro(metro_hex_id)
```
Comment (crericha) [inline lib/engine/game/g_18_usa/game.rb:1638]: Based on the other comments, I think you can make this method take in the hex object and then not have to look it up.

File: lib/engine/game/g_18_usa/step/buy_sell_par_shares.rb (hunk 1 above)
Comment (ventusignis) [reply lib/engine/game/g_18_usa/step/buy_sell_par_shares.rb:143]: thanks, changed

File: lib/engine/game/g_18_usa/step/buy_sell_par_shares.rb (hunk 3 above)
Comment (ventusignis) [reply lib/engine/game/g_18_usa/step/buy_sell_par_shares.rb:145]: Forgot visual studio sometimes refuses to save if the filesystem version changes (like if you run rubocop -A..), changed

File: lib/engine/game/g_18_usa/game.rb (hunk 2 above)
Comment (ventusignis) [reply lib/engine/game/g_18_usa/game.rb]: Changed

File: lib/engine/game/g_18_usa/game.rb (hunk 4 above)
Comment (ventusignis) [reply lib/engine/game/g_18_usa/game.rb]: Changed

**Comment by ventusignis** [comment]
> FYI I'm working on P9. Almost have it done. How do you want to coordinate?

I think slack will work for now although the message history only goes back about 3 weeks which may not work too well long term?

**Review by crericha — APPROVED** [review]

**Pull request merged**